lib/vrl/compiler/src/expression/function_call.rs

```diff
@@ -32,25 +38,28 @@ pub struct FunctionCall {
 }
 
 impl FunctionCall {
-    pub fn new(
+    #[allow(clippy::too_many_arguments)]
+    pub fn new<'a>(
         call_span: Span,
         ident: Node<Ident>,
         abort_on_error: bool,
         arguments: Vec<Node<FunctionArgument>>,
-        funcs: &[Box<dyn Function>],
-        state: &mut State,
+        closure: Option<Node<ast::FunctionClosure>>,
+        compiler: &'a mut Compiler,
```

Comment:
How do we actually want to handle our dependency on the `.compile_block` method on the Compiler? It sounds like we want to avoid passing the compiler in completely, but that method seems to be a sticking point, since the method requires a mutable borrow and we also have to mutate the compiler state.